shared/src/main/scala/tastyquery/ast/Symbols.scala

```diff
@@ -34,20 +37,56 @@ object Symbols {
 
   val NoSymbol = new RegularSymbol(nme.EmptyTermName, null)
 
-  abstract class Symbol private[Symbols] (val name: Name, rawowner: Symbol | Null) extends ParamInfo {
-
+  sealed abstract class Symbol private[Symbols] (val name: Name, rawowner: Symbol | Null) extends ParamInfo {
     private var isFlagsInitialized = false
     private var myFlags: FlagSet = Flags.EmptyFlagSet
     private var myTree: Option[DefTree] = None
-    // overridden in package symbol
-    def withTree(t: DefTree): this.type = {
+
+    /** Forces the symbol to be initialized.
+      *
+      * This method should never be called from within the initialisation of
+      * this symbol.
+      *
+      * For symbols that are not root symbols, this has no effect.
+      */
+    private[tastyquery] def ensureInitialised()(using Context): Unit
+
+    def withTree(t: DefTree): this.type =
+      require(!isPackage, s"Multiple trees correspond to one package, a single tree cannot be assigned")
       myTree = Some(t)
       this
-    }
-    def tree(using Context): Option[DefTree] = myTree
+
+    final def tree(using Context): Option[DefTree] =
+      ensureInitialised()
+      myTree
+
+    /** Get the companion class for a definition, if it exists:
+      * - for `class C` => `object class C[$]`
+      * - for `object class C[$]` => `class C`
+      * - for `object val C` => `class C`
+      */
+    final def companionClass(using Context): Option[ClassSymbol] = maybeOuter match
+      case scope: DeclaringSymbol =>
+        this match
+          case _: PackageClassSymbol => None
+          case cls: ClassSymbol =>
+            scope.getDecl(cls.name.toTypeName.companionName).collect { case sym: ClassSymbol =>
+              sym
+            }
+          case obj: RegularSymbol if obj.isTerm =>
```

Comment:
This is too broad. With a case like
```
class Foo {
  class Bar
  def Bar: Int = 5
}
```
it would link `def Bar` to `class Bar`.

We should only allow modules here.

Reply:
true, for now we don't have flags set up for modules - either I do that or temporarily require `scope` be a `PackageClassSymbol`

Reply:
I went with my suggestion for now